realm/realm-library/src/main/java/io/realm/internal/SharedGroup.java

```diff
@@ -293,6 +298,29 @@ public int hashCode() {
         }
     }
 
+    /**
+     * Waits for change committed by {@link SharedGroup} in other Thread.
+     *
+     * @return {@code true} if successfully detects change, {@code false} no change has detected otherwise.
+     */
+    public boolean waitForChange() {
```

Comment:
The concurrency guarantees are only provided by the fact that `BaseRealm.waitForChange()` has a thread check. I would be much more comfortable if this was marked `synchronized` as well. If not, as a minimum, the invariant should be documented in the Javadoc